cmake: add support for user-defined board aliases

pabigot opened the pull request:
*Inspired by the solution supporting deprecated board names for the Nordic rename, and motivated by a long standing annoyance at having to type out things like `nrf52840_pca10056` when I'm identifying the board I want to build for.*

Zephyr board names must be globally unique which requires that they
encode all necessary information to identify a specific target.
Typing in these names can be inconvenient to developers working on
multiple targets within a single workspace.

Extend the cmake infrastructure to read an optional board aliases file
that will map custom aliases to the corresponding canonical Zephyr
board name.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
cmake: add support for user-defined board aliases

Zephyr board names must be globally unique which requires that they
encode all necessary information to identify a specific target.
Typing in these names can be inconvenient to developers working on
multiple targets within a single workspace.

Extend the cmake infrastructure to read an optional board aliases file
that will map custom aliases to the corresponding canonical Zephyr
board name.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

carlescufi commented on cmake/app/boilerplate.cmake:
```diff
@@ -193,6 +193,12 @@ endif()
 assert(BOARD "BOARD not set")
 message(STATUS "Board: ${BOARD}")
 
+include(${ZEPHYR_BASE}/board_aliases.cmake OPTIONAL)
```
Wouldn't this be better placed in `boards/`?

pabigot replied:
It could be.  I prefer to keep stuff that isn't under configuration control (e.g. `hw.yml` for hardware sanitycheck testing) in the top level directory, where it's a little easier to preserve from `git clean -fdx subdir` . I'd actually prefer this to be stored out of tree somewhere, but I don't know how to do that in a way that CMake can find it.

nashif replied:
I understand the need and welcome this enhancement. However I do 'git clean -dxf' on the whole tree all the time, so this wont work, we need to keep this out of the tree to make it usefull. Look at how we store caches in user directory in cmake/extensions.cmake

pabigot replied:
@nashif `${USER_CACHE_DIR}` would be a bad location since it's undocumented and this file is expected to be maintained by the user, who may know nothing about that directory or its value.

Other options:
* create a new variable `ZEPHYR_OOT_DIR` or something less horrible?
* put it in what passes for ${HOME} and name it `.zephyr_board_aliases.cmake`
* ???

I don't like any of these alternatives, so if we can't come up with a place to put managed Zephyr customization files out of tree my next approach is to support it in west.

nashif replied:
> `${USER_CACHE_DIR}` would be a bad location since it's undocumented and this file is expected to be maintained by the user, who may know nothing about that directory or its value.

was not suggesting to use the cache dir, just use the same way we write out of the tree.

carlescufi commented on cmake/app/boilerplate.cmake:
```diff
@@ -193,6 +193,12 @@ endif()
+include(${ZEPHYR_BASE}/board_aliases.cmake OPTIONAL)
+if(${BOARD}_BOARD_ALIAS)
+  set(BOARD_ALIAS ${BOARD} CACHE STRING "Board alias, provided by user")
+  set(BOARD ${${BOARD}_BOARD_ALIAS})
```
Will this work for overlays?

pabigot replied:
If the overlay is named using the canonical name it should.  It won't find the overlay if you name it using the alias. IMO it shouldn't; the alias is just for personal convenience.

pabigot commented:
I've updated this so it now looks for an environment variable that specifies a file to be included for this purpose.

@nashif is this better?  Since this files isn't written by the Zephyr infrastructure, and for now there's only one, I don't see that we need to identify a directory to hold it.

carlescufi reviewed (approved):
Very nice, thank you.

carlescufi commented:
@tejlmand are you OK with this change?

tejlmand reviewed (changes requested):
Changes themselves looks good.

Just a small thought.
In principle, this makes it possible for a user to accidential hide an exisitng board name.
Probably not somthing that will happen for an nrf-named board, but could happen for example like this:
```
set(qemu_x86_BOARD_ALIAS qemu_x86_nommu)
```

in which case the user cannot select the real `qemu_x86`.

For sure, this is a user error, but I suggest we check the alias, to only allow using an unused board name.

Deprecated board list is different, as it is already ensured in that list, that name is not colliding with real board name as part of PR review.

tejlmand committed:
cmake: board alias message when hiding existing board

This commit will print a CMake notice if a user defines a board alias
that is identical to an existing board name.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
> > For sure, this is a user error, but I suggest we check the alias, to only allow using an unused board name.
> 
> I don't think this is a reasonable concern, and don't want to spend a lot of effort trying to cope with it. The user has complete control over her aliases. If she chooses to use a name to replace an official board, thus blocking access to it, why should that matter and why should we stop her?
> 
> But if you can you propose a _simple_ way to do such a check, and others think it's worth it, I don't strongly object to adding it.

Done:
https://github.com/pabigot/zephyr/pull/1
```
  if (BOARD_ALIAS)
    find_path(BOARD_HIDDEN_DIR
      NAMES ${BOARD_ALIAS}_defconfig
      PATHS ${root}/boards/*/*
      NO_DEFAULT_PATH
      )
    if(BOARD_HIDDEN_DIR)
      message(WARNING "Board alias ${BOARD_ALIAS} is hiding the real board of same name")
    endif()
  endif()
```

and will print:
```
Aliased BOARD=nrf51dk_nrf51422 changed to qemu_x86.
CMake Warning at /projects/github/zephyr-upstream/zephyr/cmake/app/boilerplate.cmake:302 (message):
  Board alias nrf51dk_nrf51422 is hiding the real board of same name
```

Feel free to take in the suggestion.

tejlmand reviewed (approved):
Comments addressed, and I added suggestion on how to check for aliases hiding real board names.

pabigot commented:
@tejlmand Thanks.  If these changes are not acceptable please say so, otherwise I think this is ready.

I've taken that with a couple tweaks, based on the [documentation for message modes](https://cmake.org/cmake/help/latest/command/message.html) saying that the default `NOTICE` level is for "Important message printed to stderr to attract user’s attention":

* I've downgraded the alias message to STATUS from default NOTICE, as the user created the aliases so the substitution is expected and should not be highlighted.
* I've downgraded the diagnostic on overrides from WARNING to NOTICE, since it's worth drawing attention to this but again it should be expected and not produce a warning.

Output now is:

```
tirzah[51]$ rm -rf build && west build -b frdm_k64f
-- west build: generating a build system
Including boilerplate (Zephyr base): /mnt/devel/external/zp/zephyr/cmake/app/boilerplate.cmake
-- Application: /mnt/devel/external/zp/zephyr/samples/subsys/fs/littlefs
-- Zephyr version: 2.2.99 (/mnt/devel/external/zp/zephyr)
-- Found PythonInterp: /usr/bin/python3 (found suitable version "3.7.5", minimum required is "3.6") 
-- Board: frdm_k64f
-- Aliased BOARD=frdm_k64f changed to nrf52840dk_nrf52840
Board alias frdm_k64f is hiding the real board of same name
-- Found west: /home/pab/.local/bin/west (found suitable version "0.7.1", minimum required is "0.7.1")
-- Found toolchain: zephyr (/usr/local/zephyr-sdk)
-- Found BOARD.dts: /mnt/devel/external/zp/zephyr/boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840.dts
-- Found devicetree overlay: /mnt/devel/external/zp/zephyr/samples/subsys/fs/littlefs/boards/nrf52840dk_nrf52840.overlay
...
```

tejlmand commented:
>    I've downgraded the alias message to STATUS from default NOTICE, as the user created the aliases so the substitution is expected and should not be highlighted.
>    I've downgraded the diagnostic on overrides from WARNING to NOTICE, since it's worth drawing attention to this but again it should be expected and not produce a warning.

I'm happy with that as well.
It was just a fast and simple suggestion.

So still approved.
